File: packages/python/elizaos/features/basic_capabilities/providers/choice.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


def format_choice(index: int, choice: dict[str, str]) -> str:
    label = choice.get("label", f"Option {index + 1}")
    description = choice.get("description", "")
    value = choice.get("value", str(index))

    if description:
        return f"{index + 1}. [{value}] {label}: {description}"
    return f"{index + 1}. [{value}] {label}"


async def get_choice_options(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    choices: list[dict[str, str]] = []

    if message.content and hasattr(message.content, "choices"):
        raw_choices = message.content.choices or []
        for choice in raw_choices:
            if isinstance(choice, dict):
                choices.append(choice)
            elif isinstance(choice, str):
                choices.append({"label": choice, "value": choice})

    if state and hasattr(state, "choices"):
        state_choices = state.choices or []
        for choice in state_choices:
            if isinstance(choice, dict):
                choices.append(choice)
            elif isinstance(choice, str):
                choices.append({"label": choice, "value": choice})

    if not choices:
        return ProviderResult(
            text="",
            values={"hasChoices": False, "choiceCount": 0},
            data={"choices": []},
        )

    formatted_choices = "\n".join(format_choice(i, choice) for i, choice in enumerate(choices))

    text = f"# Available Choices\n{formatted_choices}"

    return ProviderResult(
        text=text,
        values={
            "hasChoices": True,
            "choiceCount": len(choices),
            "choiceLabels": [c.get("label", "") for c in choices],
        },
        data={
            "choices": choices,
        },
    )


choice_provider = Provider(
    name="CHOICE",
    description="Available choice options for selection",
    get=get_choice_options,
    dynamic=True,
)

File: packages/rust/src/features/basic_capabilities/providers/choice.rs
//! CHOICE provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("CHOICE"));

/// Provider for choice options.
pub struct ChoiceProvider;

#[async_trait]
impl Provider for ChoiceProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        _runtime: &dyn IAgentRuntime,
        message: &Memory,
        state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        // Extract choices from message metadata
        let choices: Vec<serde_json::Value> = message
            .metadata
            .get("choices")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        // Also check state for choices
        let state_choices: Vec<serde_json::Value> = state
            .and_then(|s| s.get_value("choices"))
            .and_then(|v| v.as_array().cloned())
            .unwrap_or_default();

        let all_choices: Vec<serde_json::Value> =
            choices.into_iter().chain(state_choices).collect();

        if all_choices.is_empty() {
            return Ok(ProviderResult::new("")
                .with_value("hasChoices", false)
                .with_value("choiceCount", 0i64));
        }

        let formatted: Vec<String> = all_choices
            .iter()
            .enumerate()
            .map(|(i, choice)| {
                let label = choice
                    .get("label")
                    .and_then(|v| v.as_str())
                    .unwrap_or("Option");
                let value = choice
                    .get("value")
                    .and_then(|v| v.as_str())
                    .map(String::from)
                    .unwrap_or_else(|| i.to_string());
                format!("{}. [{}] {}", i + 1, value, label)
            })
            .collect();

        let text = format!("# Available Choices\n{}", formatted.join("\n"));

        let labels: Vec<String> = all_choices
            .iter()
            .map(|c| {
                c.get("label")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string()
            })
            .collect();

        Ok(ProviderResult::new(text)
            .with_value("hasChoices", true)
            .with_value("choiceCount", all_choices.len() as i64)
            .with_data(
                "choiceLabels",
                serde_json::to_value(&labels).unwrap_or_default(),
            )
            .with_data(
                "choices",
                serde_json::to_value(&all_choices).unwrap_or_default(),
            ))
    }
}
